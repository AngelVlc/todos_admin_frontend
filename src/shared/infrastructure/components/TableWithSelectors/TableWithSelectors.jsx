import React from "react";

export const TableWithSelectors = (props) => {
  const onSelectAll = (newCheckedValue) => {
    const newData = [...props.rows].map((row) => {
      row[props.selectedColumnName] = newCheckedValue;
      return row;
    });

    props.onSelectedChanged(newData);
  };

  const onChangeItemSelected = (newCheckedValue, index) => {
    const newData = [...props.rows];
    newData[index].selected = newCheckedValue;

    props.onSelectedChanged(newData);
  };

  return (
    <table className="table">
      <thead>
        <tr>
          {/* <td>Id
              <span class="icon">Id
                <i class="fas fa-thin fa-sort-up"></i>
              </span>
            </td> */}

          {props.columnTitles.map((columnTitle) => (
            <td key={columnTitle}>{columnTitle}</td>
          ))}
          <td>
            <center>
              <input
                type="checkbox"
                data-testid="toggleSelectAll"
                defaultChecked={false}
                onChange={(e) => onSelectAll(e.target.checked)}
              ></input>
            </center>
          </td>
        </tr>
      </thead>
      <tbody>
        {props.rows.length > 0 &&
          props.rows.map((row, index) => (
            <tr key={index}>
              {props.columnNames.map((columnName) => (
                <td key={columnName}>{row[columnName]}</td>
              ))}
              <td>
                <center>
                  <input
                    type="checkbox"
                    data-testid={`checkBoxItem${row[props.idColumnName]}`}
                    checked={row[props.selectedColumnName]}
                    onChange={(e) =>
                      onChangeItemSelected(e.target.checked, index)
                    }
                  ></input>
                </center>
              </td>
            </tr>
          ))}
      </tbody>
    </table>
  );
};
