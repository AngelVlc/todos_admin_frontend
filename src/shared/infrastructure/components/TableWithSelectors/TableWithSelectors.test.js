import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { TableWithSelectors } from "./TableWithSelectors";
import { act } from "react-dom/test-utils";

describe("TableWithSelectors", () => {
  const rows = [
    { col1: "11", col2: "12", col3: "13", selected: false },
    { col1: "21", col2: "22", col3: "23", selected: false },
  ];

  const changeSelected = jest.fn();

  const renderComponent = (rows) => {
    return {
      ...render(
        <TableWithSelectors
          columns={[
            { name: "col1", title: "Col 1" },
            { name: "col2", title: "Col 2" },
            { name: "col3", title: "Col 3" },
          ]}
          rows={rows}
          idColumnName={"col1"}
          selectedColumnName={"selected"}
          paginationInfo={{
            pageNumner: 1,
            pageSize: 10,
            sortColumn: "col1",
            sortOrder: "asc",
          }}
          changeSelected={changeSelected}
        />
      ),
    };
  };

  afterEach(cleanup);

  it("should match the snapshot", async () => {
    const { asFragment } = renderComponent(rows);

    expect(asFragment(<TableWithSelectors />)).toMatchSnapshot();
  });

  it("should select and unselect all the rows", async () => {
    let container;
    await act(async () => {
      container = renderComponent(rows);
    });

    await waitFor(() => {
      fireEvent.click(container.getByTestId("toggleSelectAll"));
    });

    expect(changeSelected.mock.calls.length).toBe(1);
    expect(changeSelected.mock.calls[0][0]).toStrictEqual([
      { col1: "11", col2: "12", col3: "13", selected: true },
      { col1: "21", col2: "22", col3: "23", selected: true },
    ]);

    changeSelected.mockClear();

    await waitFor(() => {
      fireEvent.click(container.getByTestId("toggleSelectAll"));
    });

    expect(changeSelected.mock.calls.length).toBe(1);
    expect(changeSelected.mock.calls[0][0]).toStrictEqual([
      { col1: "11", col2: "12", col3: "13", selected: false },
      { col1: "21", col2: "22", col3: "23", selected: false },
    ]);
  });

  it("should select a single row", async () => {
    let container;
    await act(async () => {
      container = renderComponent(rows);
    });

    await waitFor(() => {
      fireEvent.click(container.getByTestId("checkBoxItem11"));
    });

    expect(changeSelected.mock.calls.length).toBe(1);
    expect(changeSelected.mock.calls[0][0]).toStrictEqual([
      { col1: "11", col2: "12", col3: "13", selected: true },
      { col1: "21", col2: "22", col3: "23", selected: false },
    ]);
  });
});
