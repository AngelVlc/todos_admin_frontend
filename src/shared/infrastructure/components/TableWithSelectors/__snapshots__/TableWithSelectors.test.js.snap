// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableWithSelectors should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <table
      class="table"
    >
      <thead>
        <tr>
          <td
            data-testid="header-col1"
          >
            <span
              class="icon mx-3"
            >
              Col 1
              <i
                class="ml-1 fas fa-thin fa-sort-up"
              />
            </span>
          </td>
          <td
            data-testid="header-col2"
          >
            <span
              class="mx-3"
            >
               Col 2
            </span>
          </td>
          <td
            data-testid="header-col3"
          >
            <span
              class="mx-3"
            >
               Col 3
            </span>
          </td>
          <td>
            <center>
              <input
                data-testid="toggleSelectAll"
                type="checkbox"
              />
            </center>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            11
          </td>
          <td>
            12
          </td>
          <td>
            13
          </td>
          <td>
            <center>
              <input
                data-testid="checkBoxItem11"
                type="checkbox"
              />
            </center>
          </td>
        </tr>
        <tr>
          <td>
            21
          </td>
          <td>
            22
          </td>
          <td>
            23
          </td>
          <td>
            <center>
              <input
                data-testid="checkBoxItem21"
                type="checkbox"
              />
            </center>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="level"
    >
      <div
        class="level-left"
      >
        <div
          class="ml-3 level-item"
        >
          <input
            class="pageNumber"
            data-testid="pagenumber-input"
            min="1"
            type="number"
            value=""
          />
          <span
            class="icon"
          >
            <i
              class="fas fa-thin fa-forward"
            />
          </span>
        </div>
        <div
          class="ml-3 level-item"
        >
          <span>
            Per page: 
          </span>
          <select
            data-testid="pagesize-select"
          >
            <option
              value="5"
            >
              5
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
